### PR TITLE
Resolve race condition around exceptions during streaming a response.

### PR DESCRIPTION
**Pull request by tomchristie**
Closes https://github.com/encode/httpx/issues/2047 by resolving a race condition that can occur when an exception is raised while streaming the response.

Prior to this PR it's possible that the exception is handled first, closing the connection, and *then* the request close callback occurs.

The changes are:

* If an exception occurs while streaming a response, then close the response (and possibly the connection) 
**before allowing the exception to bubble up**.
* If a connection pool is closed while there are still open requests, then close it but then raise a `RuntimeError`.
* Defensive programming around removing the request status from the connection pool, when closing the request.

### CODE REVIEW

File: httpcore/_async/connection_pool.py
Comment (ahopkins) [inline httpcore/_async/connection_pool.py:304]: Instead of "some", it should be possible to get a number correct (`len(self._requests)`)? I know I have seen this in other exceptions or repr. I certainly appreciated that when I have seen it.

File: httpcore/_async/http11.py
Comment (ahopkins) [inline httpcore/_async/http11.py:299]: Any potential problem for setting this value before `self._connection._response_closed()`? It does not need to be set after?

Comment (ahopkins) [reply httpcore/_async/http11.py:299]: I guess if the point is to not run this multiple times it makes sense to be up front. But, thought I would ask anyway.

**Review by ahopkins — APPROVED** [review]

**Comment by tomchristie** [comment]
@ahopkins Great thanks - addressed one of your comments. Note that you'll need to accept the @encode/maintainers invitation before your review gets "merge approval" status.

**Review by ahopkins — APPROVED** [review]

**Pull request closed without merging**
